File: src/Augurk.Api/FeatureController.cs
/*
 Copyright 2014, Mark Taling
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
*/

using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Web.Http;
using Augurk.Api.Managers;
using Augurk.Entities;
using Augurk.Entities.Test;

namespace Augurk.Api
{
    public class FeatureController : ApiController
    {
        private readonly FeatureManager _featureManager = new FeatureManager();

        [Route("api/features/{branchName}")]
        [HttpGet]
        public IEnumerable<Group> Get(string branchName)
        {
            return _featureManager.GetGroupedFeatureDescriptions(branchName);
        }

        [Route("api/features/{branchName}/{groupName}/{title}")]
        [HttpGet]
        public DisplayableFeature Get(string branchName, string groupName, string title)
        {
            try
            {
                // Get the feature from storage
                DisplayableFeature feature = _featureManager.GetFeature(branchName, groupName, title);

                // Process the server tags
                if (feature != null)
                {
                    var processor = new FeatureProcessor();
                    processor.Process(feature);
                }

                return feature;
            }
            catch (Exception e)
            {
                throw e;
            }
        }

        [Route("api/features/{branchName}/{groupName}/{title}")]
        [HttpPost]
        public HttpResponseMessage Post(Feature feature, string branchName, string groupName, string title)
        {
            if (!feature.Title.Equals(title, StringComparison.OrdinalIgnoreCase))
            {
                return Request.CreateResponse(HttpStatusCode.BadRequest, "The title provided by the POST data and the title in uri do not match!");
            }

            var response = Request.CreateResponse(HttpStatusCode.Created);

            try
            {
                _featureManager.InsertOrUpdateFeature(feature, branchName, groupName);
            }
            catch (Exception exception)
            {
                response = Request.CreateErrorResponse(HttpStatusCode.InternalServerError, exception);
            }
            return response;
        }

        [Obsolete]
        [Route("api/features/{branchName}/{groupName}")]
        [HttpPost]
        public HttpResponseMessage Post(Feature feature, string branchName, string groupName)
        {
            return Post(feature, branchName, groupName, feature.Title);
        }

        [Route("api/features/{branchName}/{groupName}/{title}/testresult")]
        [HttpPost]
        public HttpResponseMessage Post(FeatureTestResult testResult, string branchName, string groupName, string title)
        {
            if (!testResult.FeatureTitle.Equals(title, StringComparison.OrdinalIgnoreCase))
            {
                return Request.CreateResponse(HttpStatusCode.BadRequest, "The title provided by the POST data and the title in uri do not match!");
            }

            var response = Request.CreateResponse(HttpStatusCode.Created);

            try
            {
                _featureManager.PersistFeatureTestResult(testResult, branchName, groupName);
            }
            catch (Exception exception)
            {
                response = Request.CreateErrorResponse(HttpStatusCode.InternalServerError, exception);
            }
            return response;
        }

        [Route("api/features/{branchName}")]
        [HttpDelete]
        public void Delete(string branchName)
        {
            _featureManager.DeleteFeatures(branchName);
        }

        [Route("api/features/{branchName}/{groupName}")]
        [HttpDelete]
        public void Delete(string branchName, string groupName)
        {
            _featureManager.DeleteFeatures(branchName, groupName);
        }

        [Route("api/features/{branchName}/{groupName}/{title}")]
        [HttpDelete]
        public void Delete(string branchName, string groupName, string title)
        {
            _featureManager.DeleteFeature(branchName, title);
        }
    }
}

File: src/Augurk/Scripts/Application/Controllers/featureController.js
/*
 Copyright 2014-2016, Mark Taling
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
*/

angular.module('Augurk').controller('featureController', ['$rootScope', '$scope', '$routeParams', 'featureService', 'featureVersionService',
    function ($rootScope, $scope, $routeParams, featureService, featureVersionService) {
        $scope.feature = featureService.get({
            productName: $routeParams.productName,
            groupName: $routeParams.groupName,
            featureName: $routeParams.featureName,
            version: $routeParams.version
        });

	    featureVersionService.versions.then(function(data) {
		    $scope.availableVersions = data;
	    });

        $rootScope.$on('currentVersionChanged', function (event, data) {
	        featureVersionService.versions.then(function(data) {
		        $scope.availableVersions = data;
	        });
        });

	    $scope.product = $routeParams.productName;
	    $scope.group = $routeParams.groupName;

        // Set the current group on the rootscope
        $rootScope.currentGroupName = $routeParams.groupName;

        // Define a function to merge the test results into the feature
        $scope.mergeTestResults = function () {
            if (!$scope.feature.testResult) {
                return;
            }

            $.each($scope.feature.testResult.scenarioTestResults, function (index, testResult) {
                $.each($scope.feature.scenarios, function (index1, scenario) {
                    //var scenario = $scope.feature.scenarios.filter(function (s) { return s.title = testResult.scenarioTitle; })[0];
                    scenario.testResult = testResult;
                });
                //var scenario = $scope.feature.scenarios.filter(function (s) { return s.title = testResult.scenarioTitle; })[0];
                //scenario.testResult = testResult;
            });

            $scope.feature.testResult.merged = true;
        };
    }
]);